### DISCUSSION TITLE
returning json - Handler<_, _> is not implemented

### DISCUSSION THREAD

**Discussion by yatesco** [tokio-rs/axum#641] [Q&A] [answered]
Hi - quite simply, why doesn't the following code work:
```rust
use axum::{Json, Router};
use axum::routing::get;
use serde_json::{json, Value};

pub(crate) fn start() -> axum::routing::Router {
    let app = Router::new()
        .route("translations.json", get(handle_locale));
    app
}

fn handle_locale() -> Json<HashMap<String,String>> {
    let m = HashMap::default();
    Json(m)
}
```

`HashMap<String,String>` is serialisable by `serde` - I don't get the problem?

The error is as follows:
```rust
error[E0277]: the trait bound `fn() -> axum::Json<HashMap<std::string::String, std::string::String>> {handle_locale}: Handler<_, _>` is not satisfied
   --> src/i18n/mod.rs:19:41
    |
19  |         .route("translations.json", get(handle_locale));
    |                                     --- ^^^^^^^^^^^^^ the trait `Handler<_, _>` is not implemented for `fn() -> axum::Json<HashMap<std::string::String, std::string::String>> {handle_locale}`
    |                                     |
    |                                     required by a bound introduced by this call
    |
note: required by a bound in `axum::routing::get`
   --> /Users/coliny/.cargo/registry/src/github.com-1ecc6299db9ec823/axum-0.4.2/src/routing/method_routing.rs:394:1
    |
394 | top_level_handler_fn!(get, GET);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `axum::routing::get`
    = note: this error originates in the macro `top_level_handler_fn` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I get the same error if I return `serde::Value` as well.

Help :-)

**Comment by jplatte** [comment, marked as answer]
If you run into `Handler<_, _> is not implemented` errors, enable axum's `macros` feature and annotate your handler with [`#[axum::debug_handler]`](https://docs.rs/axum/latest/axum/attr.debug_handler.html). It will give you much better error messages. In this case it looks like you simply forgot to make your handler function `async`.

**Reply by yatesco** [reply]
excellent - thanks!

**Reply by ScottLinnn** [reply]
Hi, I have same issue and cannot add axum_debug. It says:
```
error: no matching package found
searched package name: `axum_debug`
perhaps you meant:      axum-debug
location searched: registry `crates-io`
```
How could I handle this?

**Reply by davidpdrsn** [reply]
As stated in [the docs for axum-debug](https://docs.rs/axum-debug/latest/axum_debug) its been replaced by axum-macros.

**Reply by ScottLinnn** [reply]
Thank you!

**Reply by jplatte** [reply]
Updated my anwer to link to axum-macros instead of axum-debug 🙂

… and again to `axum`'s `macros` feature.

**Comment by hanMaster** [comment]
I have handler in my code: 
`#[axum_macros::debug_handler]
pub async fn create_account(
    Json(user): Json<RequestedUser>,
    Extension(db): Extension<DatabaseConnection>,
) -> Result<Json<ResponseUser>, StatusCode> {
  ...
}`
it cause error:
`error[E0277]: the trait bound `axum::Json<RequestedUser>: FromRequestParts<()>` is not satisfied
  --> src/routes/create_account.rs:21:17
   |
21 |     Json(user): Json<RequestedUser>,
   |                 ^^^^ the trait `FromRequestParts<()>` is not implemented for `axum::Json<RequestedUser>`
   |
`

but if I made Json as second parameter, it works fine
`#[axum_macros::debug_handler]
pub async fn create_account(
    Extension(db): Extension<DatabaseConnection>,
    Json(user): Json<RequestedUser>,
) -> Result<Json<ResponseUser>, StatusCode> {
...
}`

**Reply by jplatte** [reply]
This is intentional. https://docs.rs/axum/latest/axum/extract/index.html#the-order-of-extractors

**Reply by mustafa-ans** [reply]
@hanMaster Your solution to that error saved me lol
Thanks a million!
